In [1]:
%matplotlib inline
import math
import numpy as np
import matplotlib.pyplot as plt
import pv_model_210131 as pv


In [22]:
Iscn = 8.7 # nominal short circuit current
Vocn = 37.7 # nominal open circuit voltage
Imp = 8.2 # array current at MPP
Vmp = 30.1 # array voltage at MPP
Pmax_e = Vmp * Imp # experimental array max power output
Kv = -0.32/100 * Vocn # voltage temperature coefficient
Ki = -0.032/100 * Iscn # current temperature coefficient
Ns = 60. # number of series cells
Gn = 1000. # nominal irradiance
Tn = 25. + 273.15 # nominal operating temperature

Egap = math.pow(1.8, -19) # Bandgap of silicon (silicio cristalino)

q = 1.602 * math.pow(10, -19) # charge on electron
k = 1.3806503 * math.pow(10, -23) # boltzmann constant


In [3]:
# I_i is the value of I at the current index (if I is an array)

# non-ideal current equation
def current(Ipv, I0, I_i, V, Vt, a, Rs, Rp):
    
    I = Ipv - I0*(math.exp((V+Rs*I_i)/(Vt*a)-1)) - (V+Rs*I_i)/Rp
    
    return I

In [28]:

G = Gn 
T = Tn 
Vtn = k * Tn / q
Vt = k * T / q
deltaT = T - Tn
a = 1

Rs_max = (Vocn - Vmp) / Imp
Rp_min = Vmp / (Iscn - Imp) - Rs_max

Rp = Rp_min
Rs_vals = np.arange(0, Rs_max, 0.001)
itr = 0
v = np.arange(0.1, Vocn, 0.1)
#i = np.arange(Iscn, 0, -0.1)
i = np.zeros_like(v)

err = 0.001
p_dif = Pmax_e
Rp_prev = Rp

while p_dif >= err and itr < len(v):
    #print("outer")
    Rs = Rs_vals[itr]
    Rs_prev = Rs
    
    Ipvn = Iscn * (Rs+Rp)/Rp 
    Ipv = (Ipvn + Ki*deltaT) * G/Gn
    Isc = (Iscn + Ki*deltaT) * G/Gn
    I0n = (Ipv - Vocn/Rp) / (math.exp(Vocn/Vt/a/Ns)-1)
    I0 = I0n
    
    a = (Kv - Vocn/Tn) / (Ns * Vtn * (Ki/Ipvn - 3./Tn - Egap/(k*math.pow(Tn, 2.))))
    a = a * math.pow(10, 14)
    
    Rp_prev = Rp
    Rp = Vmp * (Vmp + Imp*Rs) / (Vmp * Ipv - Vmp* I0 * math.exp((Vmp+Imp*Rs)/Vt/Ns/a)+Vmp*I0-Pmax_e)
    
    _i = i[0]
    for idx in range(len(v)):
        _v = v[idx]
        _g = Ipv - I0 * (math.exp((_v + _i*Rs)/Vt/Ns/a)-1) - (_v + _i*Rs)/Rp - _i
        while abs(_g) > err:
            _g = Ipv - I0 * (math.exp((_v + _i*Rs)/Vt/Ns/a)-1) - (_v + _i*Rs)/Rp - _i
            _glin = -I0 * Rs/Vt/Ns/a * math.exp((_v + _i*Rs)/Vt/Ns/a) - Rs/Rp - 1
            _i = _i - _g/_glin
            i[idx] = _i
     
    P = np.zeros_like(v)
    for idx1 in range(len(v)):
        _v = v[idx1]
        _i = i[idx1]
        _p = (Ipv - I0 * (math.exp((_v + _i*Rs)/Vt/Ns/a)-1)-(_v + _i*Rs)/Rp)*_v
        P[idx1] = _p
    P_idx = np.argmax(P)
    Pmax_m = P[P_idx]
    
    p_dif = Pmax_m - Pmax_e
    itr += 1

print(Rs_prev)
print(Rp_prev, Rp)
print(Pmax_m, Pmax_e)
print(p_dif)

0.23
-260.5102369299987 -254.8331058620677
246.82 246.82
0.0


### Output

Rs = 0.371,
Rp = 586.75,
Pmax_m = 246.82036,
Pmax_e = 246.82,
p_dif = 0.00036

### Ongoing problems
a seems to be scaled by 10^-14 somewhere...